### PR TITLE
[MPS] Fix index_add with non-f32 inputs

### LINKED ISSUE #82645: [MPS] `index_select` backward error, different types are not broadcast compatible
### 🐛 Describe the bug

```python
x = torch.tensor(-1.8359375, requires_grad=True, dtype=torch.float16)
idx = torch.tensor([0])

out = torch.index_select(x, 0, idx)
out.backward()

x_mps = x.to("mps")
idx_mps = idx.to("mps")
out = torch.index_select(x_mps, 0, idx_mps)
out.backward()   # error: input types 'tensor<1xf16>' and 'tensor<1xf32>' are not broadcast compatible
```

Found in #82526 

cc @kulinseth @albanD 

### Versions

```
PyTorch version: 1.13.0a0+gitdb7187b
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] mypy==0.960
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] torch==1.13.0a0+git4eae04b
[pip3] torchvision==0.13.0
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] torch                     1.13.0a0+git4eae04b           dev_0    <develop>
[conda] torchvision               0.13.0                   pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by qqaatw**
The `multiplicationWithPrimaryTensor` and/or `scatterWithDataTensor` api has issues with handling two f16 tensor inputs, resulting in zeros outputs. With int16 or int64 inputs, there are issues as well.

This PR conditionally casts inputs to f32 if they're not and then casts the output back to the source's datatype.  

Fixes #82645.

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### CODE REVIEW

**Comment by qqaatw** [comment]
@kulinseth Hello, can you please review this one?

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 1]
```diff
@@ -354,16 +352,26 @@ Tensor flip_mps(const Tensor& self, IntArrayRef dims) {
           MPSGraphTensor* inputTensor = mpsGraphRankedPlaceHolder(mpsGraph, self);
           MPSGraphTensor* indexTensor = mpsGraphRankedPlaceHolder(mpsGraph, index);
           MPSGraphTensor* sourceTensor = mpsGraphRankedPlaceHolder(mpsGraph, source);
-          MPSGraphTensor* alphaTensor = mpsGraphScalarPlaceHolder(mpsGraph, alpha_f);
-          MPSGraphTensor* alphaSourceSlice = [mpsGraph multiplicationWithPrimaryTensor:sourceTensor
+          MPSGraphTensor* alphaTensor = mpsGraphScalarPlaceHolder(mpsGraph, getMPSScalarType(kFloat));
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Indexing.mm:355]: Hi @qqaatw, thanks for the PR. We should stick within floating types for promoting to FP32. For int types you can use i32 . And for i64 we should raise the error. (Although this is fixed in Ventura) 
src_type = isFloatingType(source.scalar_type()) ? MPSDataTypeFloat32 : MPSDataTypeInt32;

**Comment by kulinseth** [comment]
> @kulinseth Hello, can you please review this one?

Left one comment. Overall approach is good.

File: aten/src/ATen/native/mps/operations/Indexing.mm (hunk 1 above)
Comment (qqaatw) [reply aten/src/ATen/native/mps/operations/Indexing.mm]: Done.

**Comment by qqaatw** [comment]
> > @kulinseth Hello, can you please review this one?
> 
> 
> 
> Left one comment. Overall approach is good.

The suggestion has been applied, thanks.

**Comment by qqaatw** [comment]
@kulinseth Hello can you please review?

**Comment by qqaatw** [comment]
Hello @kulinseth, this PR should be good to merge.

**Comment by qqaatw** [comment]
@kulinseth Pinging for review.

**Review by kulinseth — APPROVED** [review]

**Pull request closed without merging**